### ISSUE TITLE
[docs] Disable google indexing of old docs and of master docs and of function/module summary pages and of source code listings

### ISSUE THREAD

**Issue by vadimkantorov** [pytorch/pytorch#3428]
Was googling `"pytorch" "adaptive_max_pool2d"`. Found this:
```
torch.nn — PyTorch master documentation
pytorch.org/docs/master/nn.html
... max_pool2d; max_pool3d; max_unpool1d; max_unpool2d; max_unpool3d; lp_pool2d; adaptive_max_pool1d; adaptive_max_pool2d; adaptive_max_pool3d ...
‎Torch.nn.functional · ‎Torch.optim · ‎Torch.nn.modules.loss · ‎Torch.nn.modules.rnn

torch.nn — PyTorch 0.1.12_2 documentation
pytorch.org/docs/0.1.12/nn.html
Parameters¶. class torch.nn. Parameter [source]¶. A kind of Variable that is to be considered a module parameter. Parameters are Variable subclasses, that have ...

torch.nn — PyTorch master documentation
pytorch.org/docs/0.2.0/nn.html
... max_unpool2d; max_unpool3d; lp_pool2d; adaptive_max_pool1d; adaptive_max_pool2d; adaptive_avg_pool1d; adaptive_avg_pool2d. Non-linear activation ...
```

Concerned about:
1. Very old docs from 0.1.2 indexed together with 0.2.0 and master
2. Master doc pointing to 0.2.0 for the last hit, inconsistent

This was discussed at an earlier point in https://github.com/pytorch/pytorch/issues/1124. Old docs found in Google create confusion (for OpenCV this is a very big problem, hard to find the docs of the most recent versions and to tell if the found docs are recent or not).

**Comment by vadimkantorov** [comment]
By query "pytorch mask indexing", google finds even:
```
torch — PyTorch 0.1.11_5 documentation
pytorch-zh.readthedocs.io/en/latest/torch.html
```

antique version

**Comment by vadimkantorov** [comment]
Searching for [pytorch lu factorization] returns this in top 5:
```
torch.functional — PyTorch 0.1.11_5 documentation
pytorch-zh.readthedocs.io/en/latest/_modules/torch/functional.html
def btriunpack(LU_data, LU_pivots, unpack_data=True, unpack_pivots=True): """Unpacks the data and pivots from a batched LU factorization (btrifact) of a ...
```

And another problem is that neither [pytorch lu decomposition] nor [pytorch lu factorization] returns a link to documentation of `torch.btrifact`. Maybe some bad SEO-friendliness of docs?

**Comment by zou3519** [comment]
@goodlux is working on this I believe.

**Comment by vadimkantorov** [comment]
Despite a new site, this still happens on googling "pytorch lu decomposition":

"pytorch.org/docs/0.2.0/torch.html?highlight=norm
No information is available for this page.
Learn why"

**Comment by vadimkantorov** [comment]
The adaptive_max_pool2d search still finds:
```
torch.nn — PyTorch master documentation - second hit
https://pytorch.org/docs/0.3.1/nn.html

torch.nn — PyTorch 0.1.11_5 documentation - in top 10
pytorch-zh.readthedocs.io/en/latest/nn.html
```

**Comment by soumith** [comment]
@vadimkantorov https://webmasters.stackexchange.com/questions/19216/nofollow-doesnt-stop-crawling

even after we `nofollow`, if other sites link to older docs (for example pytorch forums), then they will be ranked up

**Comment by vadimkantorov** [comment]
How about "Disallow" in robots.txt? Or setting canonical url (I think it's an og tag and also rel tag) to redirect to stable version? Also in https://github.com/pytorch/pytorch/issues/3428#issuecomment-396741517 they recommend "noindex" for those pages

The pytorch-zh is a third-party site, maintained by https://readthedocs.org/projects/pytorch-zh/, source at https://github.com/chenyuntc/pytorch-cn. I think we should try asking @chenyuntc to take down http://pytorch-zh.readthedocs.io, especially given that no docs there seems to be translated in Chinese, and the version is really old.

I think it's really important, since every user googles first before going to official docs

**Comment by chenyuntc** [comment]
https://readthedocs.org/projects/pytorch-zh/ is an effort to translate pytorch docs to Chinese since there are many people using pytorch in China. 
I've deleted them in readthedocss, sorry for the trouble.

**Comment by vadimkantorov** [comment]
@chenyuntc Thank you!

**Comment by vadimkantorov** [comment]
@soumith e.g. in robots.txt there is Disallow for 0.3.0, but no Disallow 0.3.1:
```
User-agent: *
Disallow: /docs/0.1.12/
Disallow: /docs/0.2.0/
Disallow: /docs/0.3.0/
```

**Comment by soumith** [comment]
updated robots.txt to include all of the older docs

**Comment by vadimkantorov** [comment]
@soumith it doesn’t prevent pages to be added to index (if they are linked somewhere on internet, like on pytorch forum) if I understand well (and wouldn’t remove pages already indexes). Meta noindex would remove them from index (and they must not be excluded in robots.txt)

**Comment by vadimkantorov** [comment]
Searching "pytorch grad_fn". Several comments:
1. What is `former_torchies`?
2. autograd_tutorial_old is 1.5.1. This is not old
3. autograd_tutorial is 1.0.0.dev20181128 - an old nightly!

![image](https://user-images.githubusercontent.com/1041752/87461808-338e6300-c60f-11ea-8fa3-ce3dd573376e.png)

@soumith @zou3519

**Comment by zou3519** [comment]
I think we just added versioning for the tutorials. cc @jlin27

**Comment by vadimkantorov** [comment]
@zou3519 Let's reopen this issue in the meanwhile?

**Comment by zou3519** [comment]
Yup, reopening this issue

**Comment by vadimkantorov** [comment]
Google query "pytorch jit fields" retrieves https://pytorch.org/docs/1.1.0/_modules/torch/jit.html as the first link and https://pytorch.org/docs/1.2.0/_modules/torch/jit.html as the third link:
- source code
- very old source code

Removal of source code and old docs wasn't effective 

![image](https://user-images.githubusercontent.com/1041752/90115595-9a10b900-dd54-11ea-81f1-cad35942a4f6.png)


@zou3519

**Comment by mattip** [comment]
> Google query "pytorch jit fields" retrieves ...

But googling "pytorch jit" does return the correct version. Is the goal to prevent google from __any__ indexing of older pages?

**Comment by vadimkantorov** [comment]
Yes, my proposal is to prevent google from indexing old pages altogether

**Comment by mattip** [comment]
> Google query "pytorch jit fields" retrieves https://pytorch.org/docs/1.1.0/_modules/torch/jit.html as the first link and https://pytorch.org/docs/1.2.0/_modules/torch/jit.html as the third link

Those pages have `<link rel="canonical" href="https://pytorch.org/docs/stable/_modules/torch/jit.html"/>` in the header. I wonder what else can be done?

**Comment by vadimkantorov** [comment]
I think:
1) they should be allowed under robots.txt
2) they should use noindex
for some time

After they disappear from google search, one could use robots.txt to block them from search entirely

**Comment by mattip** [comment]
So in (too much) detail:
- add a `<meta name="robots" content="noindex, nofollow, nosnippet">` line to the headers of each of the older (non-master, non-stable) pages under [docs](https://github.com/pytorch/pytorch.github.io/tree/site/docs)
- delete [this file](https://github.com/pytorch/pytorch.github.io/blob/site/robots.txt) or reduce it to have no content

**Comment by vadimkantorov** [comment]
1. master should also have these tags, since the only thing we want to have indexed is "stable". This should also apply to tutorials and everything else
2. About robots.txt - yes, I think it's not effective with Google, since if a documents leaks into Google index, then robots.txt is not helpful, but more harmful.

But I'm just a random person, have no skin in the pytorch.org game.

**Comment by zou3519** [comment]
@vadimkantorov your opinions are greatly valued!

I agree with your points that 1) only stable should be indexed and 2) robots.txt doesn't work so we should just get rid of it. Someone from the docs team (@jlin27, @brianjo) can confirm the first point.

**Comment by vadimkantorov** [comment]
google search gets very much spammed by copy-paste sites (and their bert used for retrieval isn't helpful as well) these days, especially for technical queries. so making it better on pytorch side, should make it better for many :)

**Comment by brianjo** [comment]
I'll dig into this a bit. I'll update robots.txt to exclude some of the older builds. I'll see if I can come up with something better. https://github.com/pytorch/pytorch.github.io/pull/472

**Comment by zou3519** [comment]
@brianjo adding new entries to robots.txt isn't the google recommended way of removing something from their search index. Adding `<meta name="robots" content="noindex">` tags to the html is preferred, according to https://support.google.com/webmasters/answer/93710

**Comment by brianjo** [comment]
> @brianjo adding new entries to robots.txt isn't the google recommended way of removing something from their search index. Adding `<meta name="robots" content="noindex">` tags to the html is preferred, according to https://support.google.com/webmasters/answer/93710

I would prefer not to have to touch all those old files again if possible. I'll look into other options as well.

**Comment by zou3519** [comment]
It's not the worst thing in the world to touch all the old files. For best practices, we should do a pull request per version so that if something goes wrong, we know which PR to revert.

https://github.com/pytorch/pytorch.github.io/pull/323 is an example of adding the robots tag to a large number of files. I used a sed to do all the work (the sed attempts to insert the meta robots tag directly after the `<head>` tag)
```
find . -name "*.html" -print0 | xargs -0 sed -i '/<head>/a \ \ <meta name="robots" content="noindex">'
```

**Comment by brianjo** [comment]
OK sounds good. We can plan on proceeding this way. I'll close my PR.

**Comment by mattip** [comment]
If we could commit the script to the repo, it would be one step closer to automating the process when a new version is released.

**Comment by vadimkantorov** [comment]
If the master is available in search, this can be problematic as well. e.g. like when Google ranks python 3.9.0alpha better than python 3.8

If it's existing page, maybe switch to master docs can be proposed with UX? Some overlay button?

If it's a new page, typically the API is still subject to change and only people who track the development would want to use it, and they already would know to go to master docs.

**Comment by vadimkantorov** [comment]
And also if master gets renamed into main, it will be even less clear where to click: "main" or "stable"

**Comment by vadimkantorov** [comment]
We're having part of this again, urls are correct but not the page titles. The latest stable version is 1.7.1 since two months back, but Google Search displays 1.7.0 page titles. The docs are new since two months now. Has Google not reindexed them? Do they have old titles?

Also note crappy snippets. Could some micro-formats be used to hide type signatures from search?

![image](https://user-images.githubusercontent.com/1041752/107135352-28c45400-68fa-11eb-8a4c-c468c22e4f90.png)

**Comment by brianjo** [comment]
Thanks for letting me know. @zou3519 could you run your magic script on the https://github.com/pytorch/pytorch.github.io/tree/site/docs/1.7.1 folder?

**Comment by zou3519** [comment]
@vadimkantorov the problem looks like either:
- the 1.7.1 documentation have page titles that erroneously say 1.7.0.
- the 1.7.1 documentation is actually the 1.7.0 documentation

If we go to https://pytorch.org/docs/stable/torch.html, the page actually says 
![image](https://user-images.githubusercontent.com/5652049/107248107-fac93780-69ff-11eb-81d0-445df19e2e0c.png)


@brianjo the magic script doesn't solve this problem -- the magic script removes a version of the documentation from google indexing.

**Comment by mattip** [comment]
So there are at least two problems
- somehow the page `<title>` for the 1.7.1 docs (in the top few lines of the HTML) says 1.7.0. This is easily fixed with a sed replacement script, It would be good to figure out how this happened and prevent it in the future. The 1.8.0 docs were generated (I think) correctly. The original version of the documents in commit `pytorch/pyotrch.github.io/c403592bb` from Nov 22 already had the 1.7.0 title text.
- it would be nice to figure out how to hide the signature from the (arguably more appropriate) generated page when the summary is shown in search. Is there a way to do that?

**Comment by vadimkantorov** [comment]
> * it would be nice to figure out how to hide the signature from the (arguably more appropriate) generated page when the summary is shown in search. Is there a way to do that?

Maybe it can be done by having different values in title tag and in meta tag. But as long as the version value is correct and most recent, maybe it's fine?

**Comment by vadimkantorov** [comment]
Master documentation is again in Google Index: 
![image](https://user-images.githubusercontent.com/1041752/115714701-b64f0600-a377-11eb-9ed4-4c681fe6978f.png)

@zou3519 could we disable it somehow?

**Comment by zou3519** [comment]
> Master documentation is again in Google Index.
@zou3519 could we disable it somehow?

I don't think we ever came to a decision on if master documentation should be searchable by google or not. My inclination is for it not to be searchable because we expect users to land on the stable pages of PyTorch. What are your thoughts, @brianjo @mattip?

**Comment by brianjo** [comment]
Yeah that's a good idea. Can we run our add_noindex_tags.sh script as part of the nightly build?

**Comment by vadimkantorov** [comment]
Are function summary pages available for google search? I propose to remove them from the search either (if they're not already excluded)

**Comment by mattip** [comment]
Do we need to? If I search for [pytorch std](https://www.google.com/search?hl=en&q=pytorch%20std) the summary page "torch — PyTorch 1.8.1 documentation" is quite far down

**Comment by vadimkantorov** [comment]
I think it would help avoiding confusion, so in general a good thing IMO.

Also IMO the next big battle should be for good search snippets / captions - I propose excluding code examples from snippets and maybe function arguments as well. Currently snippets often are a unreadable mixture of parts of the signature and parts of functionality description - the worst

**Comment by vadimkantorov** [comment]
Also, master docs are still present in Google search despite the commit 8 days ago

**Comment by zou3519** [comment]
@vadimkantorov do you have an example? I haven't been seeing master docs:
![image](https://user-images.githubusercontent.com/5652049/119706633-3dd0de80-be28-11eb-82ec-d965a1c85b76.png)

Maybe it takes time for google to process all of the noindex tags. If I try really hard I can find some master doc pages
![image](https://user-images.githubusercontent.com/5652049/119706734-580abc80-be28-11eb-8eb2-9e2e01df821e.png)

**Comment by vadimkantorov** [comment]
> Do we need to? If I search for [pytorch std](https://www.google.com/search?hl=en&q=pytorch%20std) the summary page "torch — PyTorch 1.8.1 documentation" is quite far down

I had them for this query

**Comment by zou3519** [comment]
Cool, I'm going to wait a week or so and check back in on if google decided to remove those. Regarding the other two points:

> Also IMO the next big battle should be for good search snippets / captions - I propose excluding code examples from snippets and maybe function arguments as well. Currently snippets often are a unreadable mixture of parts of the signature and parts of functionality description

> Are function summary pages available for google search? I propose to remove them from the search as well (if they're not already excluded)

Let's make separate issues for each of those instead of turning this issue into a general "improve docs seo tracking issue".

**Comment by vadimkantorov** [comment]
For snippets, let's figure out first if Google has any microformats for allowing that...

**Comment by vadimkantorov** [comment]
Any idea where the bold word in the snippet came from?
![image](https://user-images.githubusercontent.com/1041752/120210080-aa384d00-c22f-11eb-96b8-47b334be4f90.png)

**Comment by mattip** [comment]
I guess that is the way google formats the table at the top of [this page](https://pytorch.org/docs/master/nn.init.html), whose source is https://pytorch.org/docs/master/_sources/nn.init.rst.txt. That page could be broken up like [the PRs I did](https://github.com/pytorch/pytorch/pull/55748/files) for the larger pages by replacing `autofunction` with `autosummary`. Also note that the actual function `trunc_normal_`  should be added to that list.

**Comment by vadimkantorov** [comment]
torchvision still lets Google index old docs @fmassa:
![image](https://user-images.githubusercontent.com/1041752/126240732-7fd56102-b12c-4823-869d-30b40d1a8828.png)

**Comment by vadimkantorov** [comment]
Summary page for functions is ranked higher than individual function page. It would be good to de-index summary pages:
![image](https://user-images.githubusercontent.com/1041752/139822320-f67ba0d6-8b0c-4e35-a04d-ca4f8aedc41d.png)

**Comment by svekars** [comment]
Sorry, I missed this ping. We add `<meta name="robots" content="noindex">` to previous version of the docs a on every release and we have the same for the main docs as well. I believe it achieves what we need, no?

**Comment by zou3519** [comment]
Yeah I think we have done this
